File: Algorithms/049. Group Anagrams/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        map={}
        strs.sort()
        for s in strs:
            l=list(s)
            l.sort()
            eigenValue=str(l)
            if map.get(eigenValue) is None:
                map[eigenValue]=[s]
            else:
                map.get(eigenValue).append(s)
        return map.values()

File: Algorithms/049. Group Anagrams/Solution.java
public class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Arrays.sort(strs);
        String[] newStrs=new String[strs.length];
        for (int i = 0; i < strs.length; i++) {
            char[] c=strs[i].toCharArray();
            Arrays.sort(c);
            newStrs[i]=String.valueOf(c);
        }
        Map<String,List<Integer>> map=new HashMap<String, List<Integer>>();
        for (int i = 0; i < newStrs.length; i++) {
            String str=newStrs[i];
            List<Integer> list=null;
            if (map.get(str)==null) {
                list=new ArrayList<Integer>();
            }else {
                list=map.get(str);
            }
            list.add(i);
            map.put(str, list);
        }
        List<List<String>> ans=new ArrayList<List<String>>();
        Set<String> keySet = map.keySet();
        for (String s:keySet) {
            List<Integer> l=map.get(s);
            List<String> list=new ArrayList<String>();
            for (int i = 0; i < l.size(); i++) {
                list.add(strs[l.get(i)]);
            }
            ans.add(list);
        }
        return ans;
    }
}